temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientInvoker.java

```diff
@@ -356,20 +349,110 @@ public <R> UpdateOutput<R> update(UpdateInput<R> input) {
+            .build();
+
+    PollWorkflowExecutionUpdateRequest pollUpdateRequest =
+        PollWorkflowExecutionUpdateRequest.newBuilder()
+            .setNamespace(clientOptions.getNamespace())
+            .setIdentity(clientOptions.getIdentity())
+            .setUpdateRef(update)
+            .setWaitPolicy(waitPolicy)
+            .build();
+
+    CompletableFuture<PollWorkflowExecutionUpdateResponse> future = new CompletableFuture<>();
+
+    Deadline pollTimeoutDeadline = Deadline.after(input.getTimeout(), input.getTimeoutUnit());
+    pollWorkflowUpdateHelper(future, pollUpdateRequest, pollTimeoutDeadline);
+    return new PollWorkflowUpdateOutput(
+        future.thenApply(
+            (result) -> {
+              if (result.hasOutcome()) {
+                switch (result.getOutcome().getValueCase()) {
+                  case SUCCESS:
+                    Optional<Payloads> updateResult = Optional.of(result.getOutcome().getSuccess());
+                    R resultValue =
+                        convertResultPayloads(
+                            updateResult,
+                            input.getResultClass(),
+                            input.getResultType(),
+                            dataConverterWithWorkflowContext);
+                    return resultValue;
+                  case FAILURE:
+                    throw new WorkflowUpdateException(
+                        input.getWorkflowExecution(),
+                        input.getUpdateId(),
+                        input.getUpdateName(),
+                        dataConverterWithWorkflowContext.failureToException(
+                            result.getOutcome().getFailure()));
+                  default:
+                    throw new RuntimeException(
+                        "Received unexpected outcome from poll update request: "
+                            + result.getOutcome().getValueCase());
+                }
+              }
+              throw new RuntimeException("Received no outcome from server");
+            }));
+  }
+
+  private void pollWorkflowUpdateHelper(
+      CompletableFuture<PollWorkflowExecutionUpdateResponse> resultCF,
+      PollWorkflowExecutionUpdateRequest request,
+      Deadline deadline) {
+
+    Deadline pollTimeoutDeadline = Deadline.after(60L, TimeUnit.SECONDS).minimum(deadline);
+    genericClient
+        .pollUpdateAsync(request, pollTimeoutDeadline)
+        .whenComplete(
+            (r, e) -> {
+              if ((e instanceof StatusRuntimeException
+                      && ((StatusRuntimeException) e).getStatus().getCode()
+                          == Status.Code.DEADLINE_EXCEEDED)
+                  || pollTimeoutDeadline.isExpired()
+                  || (e == null && !r.hasOutcome())) {
+                // if the request has timed out, stop retrying
+                if (!deadline.isExpired()) {
+                  pollWorkflowUpdateHelper(resultCF, request, deadline);
```

Comment:
Is `deadline` clamped somewhere? Obviously there's a stack depth concern here but let's say someone wanted to wait for 24h, with 1m polls we're only looking at something on the order of 1440 frames so ... maybe ok? I guess we're prevented from using an iterative approach because of whenComplete?

Reply:
This is actually not recursive. `pollWorkflowUpdateHelper`  will return immediately , it just launches another poll request that will write to the result future.